airflow/traces/otel_tracer.py

```diff
@@ -0,0 +1,333 @@
+        return span
+
+    def start_span_from_taskinstance(
+        self,
+        ti,
+        span_name: str | None = None,
+        component: str = "taskinstance",
+        child: bool = False,
+        links=None,
+    ):
+        """
+        Create and start span from given task instance.
+
+        Essentially the span represents the ti itself if child == True, it will create a 'child' span under the given span.
+        """
+        dagrun = ti.dag_run
+        trace_id = int(gen_trace_id(dag_run=dagrun), 16)
+        span_id = int(gen_span_id(ti=ti), 16)
+        if span_name is None:
+            span_name = ti.task_id
+
+        if child is False:
+            parent_id = int(gen_dag_span_id(dag_run=dagrun), 16)
+        else:
+            parent_id = span_id
+
+        span_ctx = SpanContext(
+            trace_id=trace_id, span_id=parent_id, is_remote=True, trace_flags=TraceFlags(0x01)
+        )
+
+        if links is not None:
+            _links = gen_links_from_kv_list(links)
+        else:
+            _links = []
+
+        a_link = Link(
+            context=trace.get_current_span().get_span_context(),
+            attributes={"meta.annotation_type": "link", "from": "parenttrace"},
+        )
+        _links.append(a_link)
+
+        if child is False:
+            tracer = self.get_tracer_with_id(component=component, span_id=span_id, trace_id=trace_id)
+        else:
+            tracer = self.get_tracer(component=component)
+
+        ctx = trace.set_span_in_context(NonRecordingSpan(span_ctx))
+        span = tracer.start_as_current_span(
+            name=span_name, context=ctx, start_time=int(ti.queued_dttm.timestamp() * 1000000000), links=_links
+        )
+        return span
+
+
+def gen_context(trace_id: int, span_id: int):
+    """Generate a remote span context for given trace and span id."""
+    span_ctx = SpanContext(trace_id=trace_id, span_id=span_id, is_remote=True, trace_flags=TraceFlags(0x01))
+    return span_ctx
+
+
+def gen_links_from_kv_list(kv_list):
+    """Convert list of kv dic of trace_id and span_id and generate list of SpanContext."""
+    result = []
+    for a in kv_list:
+        trace_id = a["trace_id"]  # string of hexa
+        span_id = a["span_id"]  # string of hexa
+        span_ctx = gen_context(trace_id, span_id)
+        a_link = Link(
+            context=span_ctx,
+            attributes={"meta.annotation_type": "link"},
+        )
+        result.append(a_link)
+    return result
+
+
+def gen_link_from_traceparent(traceparent: str):
+    """Generate Link object from provided traceparent string."""
+    if traceparent is not None:
+        trace_ctx = parse_traceparent(traceparent)
+        trace_id = trace_ctx["trace_id"]
+        span_id = trace_ctx["parent_id"]
+        span_ctx = gen_context(int(trace_id, 16), int(span_id, 16))
+        a_link = Link(context=span_ctx, attributes={"meta.annotation_type": "link", "from": "traceparent"})
+        return a_link
+    return None
```

Comment:
This may be a personal preference thing, but I was taught to put your escape hatch early, so this method would be:


```suggestion
def gen_link_from_traceparent(traceparent: str):
    """Generate Link object from provided traceparent string."""
    if traceparent is None:
        return None
        
    trace_ctx = parse_traceparent(traceparent)
    trace_id = trace_ctx["trace_id"]
    span_id = trace_ctx["parent_id"]
    span_ctx = gen_context(int(trace_id, 16), int(span_id, 16))
    
    return Link(context=span_ctx, attributes={"meta.annotation_type": "link", "from": "traceparent"})
```

instead of indenting the whole method

Reply:
Yes, escaping early does make sense, and it also makes the code much more readable. Thanks!